containers/src/Kokkos_Vector.hpp

```diff
@@ -147,6 +147,83 @@ class vector : public DualView<Scalar*, LayoutLeft, Arg1Type> {
 
   void clear() { _size = 0; }
 
+  iterator insert(iterator it, const value_type& val) {
+    return insert(it, 1, val);
+  }
+
+  iterator insert(iterator it, size_type count, const value_type& val) {
+    if ((_size == 0) && (it == begin())) {
+      resize(count, val);
+      DV::sync_host();
+      return begin();
+    }
+    DV::sync_host();
+    DV::modify_host();
+    if (it < begin() || it > end())
+      Kokkos::abort("Kokkos::vector::insert : invalid insert iterator");
+    if (count == 0) return it;
+    ptrdiff_t start = std::distance(begin(), it);
+    auto org_size   = size();
+    resize(size() + count);
+    auto org_view =
+        subview(DV::h_view, pair<ptrdiff_t, ptrdiff_t>(start, org_size));
+    auto new_view = subview(DV::h_view, pair<ptrdiff_t, ptrdiff_t>(
+                                            start + count, org_size + count));
+
+    for (ptrdiff_t i = org_view.extent(0) - 1; i >= 0; i--) {
+      new_view(i) = org_view(i);
+    }
+
+    for (ptrdiff_t i = start; i < start + (ptrdiff_t)count; i++)
+      DV::h_view(i) = val;
+    return begin() + start;
+  }
+
+ private:
+  template <class T>
+  struct impl_is_input_iterator
+      : /* TODO replace this */ std::integral_constant<
+            bool, !std::is_convertible<T, size_type>::value> {};
+
+ public:
+  // TODO: can use detection idiom to generate better error message here later
+  template <typename InputIterator>
+  typename std::enable_if<impl_is_input_iterator<InputIterator>::value,
+                          iterator>::type
+  insert(iterator it, InputIterator b, InputIterator e) {
+    ptrdiff_t count = std::distance(b, e);
+    if (count == 0) return it;
+
+    if ((_size == 0) && (it == begin())) {
+      resize(count);
+      it = begin();
+    }
+    DV::sync_host();
+    DV::modify_host();
+    if (it < begin() || it > end())
+      Kokkos::abort("Kokkos::vector::insert : invalid insert iterator");
+
+    ptrdiff_t start = it - begin();
+    auto org_size   = size();
+    resize(size() + count);
```

Comment:
Here. You resize the formerly empty vector a 2nd time.

Reply:
And is that a problem? Are you arguing this doesn't test iterator invalidation? I think I found what went wrong with my Intel build. I believe the GCC used with Intel 16 turned out to be GCC 4.4 and thus that particular version of std::vector included was still a C++98 version which on that overload of insert returns void.

Reply:
I am just showing you where the 2nd resize would occur in https://github.com/kokkos/kokkos/pull/2443#discussion_r334080343